services/endpointsrv/service.go

```diff
@@ -70,105 +103,97 @@ func NewEndpointService(
 	}
 }
 
+// EndpointExecCreate creates a new endpoint (with http address)
+// it also creates the first endpoint version for the exec
+// promotes that version to primary, and serves traffic to it.
 func (e *EndpointService) EndpointExecCreate(ctx context.Context, projectID, execID, endpointName string) (types.Endpoint, error) {
 	endpointID := typeid.Must(typeid.New("end")).String()
 
 	if endpointName == "" {
 		endpointName = random.GenerateRandomPhrase(3, "-")
 	}
 
-	subdomain := fmt.Sprint(endpointName, "-", random.GenerateRandomLower(5))
-
-	exec, err := e.execs.Get(ctx, execID)
+	end, err := e.createEndpoint(ctx, endpointID, endpointName, projectID)
 	if err != nil {
-		return types.Endpoint{}, fmt.Errorf("get exec: %w", err)
-	}
-
-	if exec.Provider != e.driver.EndpointProvider() {
-		return types.Endpoint{}, &types.Error{
-			Message:    "Cannot create endpoint",
-			Suggestion: fmt.Sprintf("Endpoints can only be created with the %s provider", e.driver.EndpointProvider().String()),
-			Provider:   e.driver.EndpointProvider(),
-		}
+		return types.Endpoint{}, fmt.Errorf("create endpoint: %w", err)
 	}
 
-	httpSerivice := exec.Network.HTTPService
+	return end, nil
+}
 
-	if httpSerivice == nil {
-		return types.Endpoint{}, &types.Error{
-			Code:       400,
-			Message:    "Endpoints can only be created on execs that expose a service",
-			Suggestion: "Try creating an exec with a port exposed",
-		}
-	}
+func (e *EndpointService) createEndpoint(
+	ctx context.Context,
+	endpointID,
+	endpointName,
+	projectID string,
+) (types.Endpoint, error) {
+	subdomain := fmt.Sprint(endpointName, "-", random.GenerateRandomLower(5))
 
 	httpAddr, err := e.driver.EndpointCreate(
 		ctx,
 		projectID,
 		endpointID,
-		execID,
 		subdomain,
-		httpSerivice.InternalPort,
+		8080,
```

Comment:
Port is always 8080? Is this a regression or on purpose?

Reply:
Yes, so this is tricky. 

- An `Endpoint` points at a `Version`
- A `Version` points at an `Exec`
- An `Exec` has a port. 

Purely for the networking/connections an Endpoint is just a nice-name over a specific "primary" version. The Version & Exec know the ports, and the Endpoint just forwards to that. 

In concrete terms, it could be something like: 
- Endpoint: CNAME
- Version: Loadbalancer (+port) & DNS name
- Exec: VM (+port)

So arguably we shouldn't need a port in the interface/abstraction for creating an endpoint :thinking: I'll take it out.